gno.land/pkg/sdk/vm/keeper.go

```
@@ -136,6 +140,102 @@ func (vm *VMKeeper) getGnoStore(ctx sdk.Context) gno.Store {
+	const sysUsersPkg = "gno.land/r/demo/users"
+
+	store := vm.getGnoStore(ctx)
+
+	// if `sysUsersPkg` does not exists -> skip validation.
+	usersPkg := store.GetPackage(sysUsersPkg, false)
+	if usersPkg == nil {
+		return nil
+	}
+
+	// XXX: is this necessary ?
+	pkgPath = path.Clean(pkgPath) // cleanup pkgpath
+
+	match := reNamespace.FindStringSubmatch(pkgPath)
+	if len(match) != 2 {
+		return ErrInvalidPkgPath(pkgPath)
+	}
+	username := match[1]
+
+	// Lowercase username
+	username = strings.ToLower(username)
+
+	// Allow user with their own address as namespace
+	if addr, err := crypto.AddressFromBech32(username); err == nil {
+		if addr.Compare(creator) == 0 {
+			return nil // ok
+		}
+
+		return ErrUnauthorizedUser(username)
+	}
+
+	// Parse and run the files, construct *PV.
+	pkgAddr := gno.DerivePkgAddr(pkgPath)
+	msgCtx := stdlibs.ExecContext{
+		ChainID:       ctx.ChainID(),
+		Height:        ctx.BlockHeight(),
+		Timestamp:     ctx.BlockTime().Unix(),
+		OrigCaller:    creator.Bech32(),
+		OrigSendSpent: new(std.Coins),
+		OrigPkgAddr:   pkgAddr.Bech32(),
+		// XXX: should we remove the banker ?
+		Banker:      NewSDKBanker(vm, ctx),
+		EventLogger: ctx.EventLogger(),
+	}
+
+	m := gno.NewMachineWithOptions(
+		gno.MachineOptions{
+			PkgPath:   "",
+			Output:    os.Stdout, // XXX
+			Store:     store,
+			Context:   msgCtx,
+			Alloc:     store.GetAllocator(),
+			MaxCycles: vm.maxCycles,
+			GasMeter:  ctx.GasMeter(),
+		})
+	defer m.Release()
+
+	// call $sysUsersPkg.Resolve("<user>")
+	// We only need to check by name here, as address have already been check
+	mpv := gno.NewPackageNode("main", "main", nil).NewPackage()
+	m.SetActivePackage(mpv)
+	m.RunDeclaration(gno.ImportD("users", sysUsersPkg))
+	x := gno.Call(
+		gno.Sel(gno.Nx("users"), "GetUserAddressByName"),
+		gno.Str(username),
+	)
+
+	ret := m.Eval(x)
+	if len(ret) == 0 {
+		panic("call: invalid response length")
```

Why not return an error?

---

This should not happen, there should be no ambiguity.